metricsx/bridge_segment.go

```diff
@@ -0,0 +1,238 @@
+package metricsx
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net"
+	"net/http"
+	"runtime"
+	"time"
+
+	"github.com/pborman/uuid"
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+	"github.com/segmentio/analytics-go"
+	"github.com/sirupsen/logrus"
+
+	"github.com/ory/x/resilience"
+)
+
+type analyticsProperties struct {
+	memstats analytics.Properties
+	app      analytics.Properties
+}
+
+// This map is for converting Prometheus metrics (by name) to segment keys that were
+// previously sent from ory/x/metricsx
+var memStats = map[string]string{
+	"go_memstats_heap_alloc_bytes":    "alloc",
+	"go_memstats_alloc_bytes_total":   "totalAlloc",
+	"go_memstats_sys_bytes":           "sys",
+	"go_memstats_lookups_total":       "lookups",
+	"go_memstats_mallocs_total":       "mallocs",
+	"go_memstats_frees_total":         "frees",
+	"go_memstats_heap_sys_bytes":      "heapSys",
+	"go_memstats_heap_idle_bytes":     "heapIdle",
+	"go_memstats_heap_inuse_bytes":    "heapInuse",
+	"go_memstats_heap_released_bytes": "heapReleased",
+	"go_memstats_heap_objects":        "heapObjects",
+	"go_gc_duration_seconds_count":    "numGC",
+}
+
+func labelsContain(key string, labels []*dto.LabelPair) bool {
+	for _, v := range labels {
+		if v.GetName() == key {
+			return true
+		}
+	}
+	return false
+}
+
+func getSegmentKey(mf *dto.MetricFamily) (string, bool) {
+	if val, ok := memStats[mf.GetName()]; ok {
+		return val, true
+	}
+	return mf.GetName(), false
+}
+
+func getValueFromMetric(t dto.MetricType, m *dto.Metric) float64 {
+	switch t {
+	case dto.MetricType_GAUGE:
+		return m.GetGauge().GetValue()
+	case dto.MetricType_COUNTER:
+		return m.GetCounter().GetValue()
+	}
+	return 0
+}
+
+func getSegmentProperties(mfs []*dto.MetricFamily) analyticsProperties {
+	p := analyticsProperties{
+		memstats: analytics.NewProperties(),
+		app:      analytics.NewProperties(),
+	}
+	for _, v := range mfs {
+		for _, m := range v.GetMetric() {
+			if v.GetType() == dto.MetricType_HISTOGRAM || v.GetType() == dto.MetricType_SUMMARY {
+				// Don't transfer histogram buckets / summary percentiles to segment
```

Comment:
Is there a way to ensure that we only send stuff to segment that segment actually understands? Are there more types than the listed ones maybe?

Reply:
I've come up with a much better solution to this. :)